Grouped product frontend quantity validation added and code refactor

Mohamed-Asar opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
Grouped Product Frontend Quantity validation added

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
1. #39479 
2. #39142

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->

1.Create a Grouped Product and associate it with child products.
2. Set the following quantity properties for one of the child products:
    - Minimum Qty Allowed in Shopping Cart: 2
    - Maximum Qty Allowed in Shopping Cart: 10
    - Qty Increments: 2
 3.Open the Product Detail Page (PDP) of the Grouped Product and test invalid quantities for the selected child product (e.g., 1, 3, or 11) to confirm that invalid inputs are restricted.



### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

Mohamed-Asar committed:
Grouped product frontend quantity validation added and code refactor

Mohamed-Asar committed:
Merge branch '2.4-develop' into bug/39479-grouped-product-qty-validation

Mohamed-Asar commented:
@magento run all tests.

Mohamed-Asar commented:
@magento run all tests

Mohamed-Asar committed:
Merge branch '2.4-develop' of https://github.com/Mohamed-Asar/magento2 into bug/39479-grouped-product-qty-validation

Mohamed-Asar committed:
Ignore item qty validation if zero qty entered in grouped product

Mohamed-Asar committed:
Merge branch 'bug/39479-grouped-product-qty-validation' of https://github.com/Mohamed-Asar/magento2 into bug/39479-grouped-product-qty-validation

Mohamed-Asar commented:
@magento run all tests

Mohamed-Asar committed:
Merge branch '2.4-develop' of https://github.com/Mohamed-Asar/magento2 into bug/39479-grouped-product-qty-validation

Mohamed-Asar committed:
Unit test & static test cases issues fixed

Mohamed-Asar commented:
@magento run all tests

Mohamed-Asar committed:
Static test fixed

Mohamed-Asar committed:
Indent issue fixed

Mohamed-Asar commented:
@magento run all tests

engcom-Hotel committed:
Merge branch '2.4-develop' into bug/39479-grouped-product-qty-validation

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented on app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php:
```diff
@@ -0,0 +1,53 @@
+
+namespace Magento\CatalogInventory\Model\Product;
+
+use Magento\CatalogInventory\Api\StockRegistryInterface;
```
interface StockRegistryInterface has been deprecated. Please use alternate.

Mohamed-Asar replied:
@engcom-Hotel The alternative for that interface is the MSI module. However, some users might have the MSI module disabled, can i use that ?

engcom-Hotel replied:
Yes @Mohamed-Asar you can use that. BTW its not recommended to disable the MSI module.

Mohamed-Asar replied:
@engcom-Hotel MSI module is not included in this repo. how do i proceed ?

engcom-Hotel replied:
Please refer to this repo: https://github.com/magento/inventory

Mohamed-Asar replied:
@engcom-Hotel I used MSI and remvoed the deprecated methods, please review

engcom-Hotel commented on app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php:
```diff
@@ -0,0 +1,53 @@
+     * @param StockRegistryInterface $stockRegistry
+     */
+    public function __construct(
+        StockRegistryInterface $stockRegistry
```
Use `readonly` instead. Please refer to [this](https://github.com/magento/magento2/blob/3428ebc3802a26fae9be542163579cefaa7967fa/app/code/Magento/Tax/Model/Plugin/AddTaxesExtensionAttribute.php#L46).

Mohamed-Asar replied:
done.

engcom-Hotel commented on app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php:
```diff
@@ -0,0 +1,53 @@
+     *
+     * @return array
+     */
+    public function getData($productId, $websiteId): array
```
The parameter's type declaration is missing. Please add.

Mohamed-Asar replied:
done

engcom-Hotel commented on app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php:
```diff
@@ -0,0 +1,53 @@
+
+        $params = [];
+        $validators = [];
+        $params['minAllowed'] = (float)$stockItem->getMinSaleQty();
```
`getMinSaleQty` is returning the `float` value. No need to typecast here.

Mohamed-Asar replied:
removed the typecast

engcom-Hotel commented on app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php:
```diff
@@ -0,0 +1,53 @@
+        $validators = [];
+        $params['minAllowed'] = (float)$stockItem->getMinSaleQty();
+        if ($stockItem->getMaxSaleQty()) {
+            $params['maxAllowed'] = (float)$stockItem->getMaxSaleQty();
```
Same here

Mohamed-Asar replied:
removed the typecast

engcom-Hotel commented on app/code/Magento/GroupedProduct/ViewModel/ValidateQuantity.php:
```diff
@@ -0,0 +1,58 @@
+     * @param QuantityValidator $productQuantityValidator
+     */
+    public function __construct(
+        Json $serializer,
```
Use `readonly` constructor parameter.

Mohamed-Asar replied:
done

engcom-Hotel commented on app/code/Magento/GroupedProduct/ViewModel/ValidateQuantity.php:
```diff
@@ -0,0 +1,58 @@
+     *
+     * @return string
+     */
+    public function getQuantityValidators($productId, $websiteId): string
```
Please add parameter's type declaration.

Mohamed-Asar replied:
done

engcom-Hotel commented on app/code/Magento/CatalogInventory/Block/Plugin/ProductView.php:
```diff
@@ -1,26 +1,26 @@
      */
     public function __construct(
-        StockRegistryInterface $stockRegistry
+        QuantityValidator $productQuantityValidator
```
Use ObjectManager to initialize as this change is BiC.

Mohamed-Asar replied:
done

engcom-Hotel replied:
not done yet, still cant see ObjectManager usages

Mohamed-Asar replied:
@engcom-Hotel i tried that, but the static test failed due to the object being initialized as null, so I had to revert the changes.

![image](https://github.com/user-attachments/assets/e42770c4-e34c-43c5-b45b-478d798174b7)

engcom-Hotel commented:
@magento run Functional Tests B2B, Functional Tests EE, Functional Tests CE, Static Tests, Unit Tests

engcom-Hotel reviewed (changes requested):
Hello @Mohamed-Asar,

Thanks for the contribution!

Please find below the review comments.

Thanks

Mohamed-Asar committed:
Code review changes

engcom-Hotel commented:
Moving back to `Changes requested` as 1 [review comment](https://github.com/magento/magento2/pull/39480#discussion_r1926716718) still needs to be worked on.

engcom-Hotel commented:
@magento run all tests

Mohamed-Asar commented:
@engcom-Hotel  Not able to install MSI module, facing the below issue 

![image](https://github.com/user-attachments/assets/095c6de7-7658-42a4-8e6e-d6df2e12fa8b)

This issue occurs due to recent changes, which was already discussed [here](https://magentocommeng.slack.com/archives/C7KB93M32/p1736933306053669), i will proceed once the issue resolved

engcom-Hotel commented:
Sure @Mohamed-Asar, until we are moving the PR `On Hold`.

Thanks

Mohamed-Asar committed:
Merge branch '2.4-develop' of https://github.com/Mohamed-Asar/magento2 into bug/39479-grouped-product-qty-validation

Mohamed-Asar commented:
> @engcom-Hotel Not able to install MSI module, facing the below issue
> 
> ![image](https://private-user-images.githubusercontent.com/56036656/408021324-095c6de7-7658-42a4-8e6e-d6df2e12fa8b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDExNjAwOTksIm5iZiI6MTc0MTE1OTc5OSwicGF0aCI6Ii81NjAzNjY1Ni80MDgwMjEzMjQtMDk1YzZkZTctNzY1OC00MmE0LThlNmUtZDZkZjJlMTJmYThiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzA1VDA3Mjk1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMwMTY1MmMyYjAzMDEwODk5NTliMGRhNWFmZTQ5YTQ4ODA5MDIxY2NkMWEwN2E2ZDQzMDFjZTI4Mjg2NjdhNzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.pBQEk4_yozeFulXYVPw_SSN2n20u96kq60zbNo_s8DI)
> 
> This issue occurs due to recent changes, which was already discussed [here](https://magentocommeng.slack.com/archives/C7KB93M32/p1736933306053669), i will proceed once the issue resolved

@engcom-Hotel This issue was fixed by internal team, i will start working on the changes and let you know

Mohamed-Asar committed:
Merge branch '2.4-develop' of https://github.com/Mohamed-Asar/magento2 into bug/39479-grouped-product-qty-validation

Mohamed-Asar committed:
Removed deprecated methods, get data using MSI and unit test added

Mohamed-Asar commented:
@magento run all tests

Mohamed-Asar committed:
Static test case fix

engcom-Hotel committed:
Merge branch '2.4-develop' into bug/39479-grouped-product-qty-validation

engcom-Hotel commented:
@magento run all tests

engcom-Hotel committed:
Merge branch '2.4-develop' into bug/39479-grouped-product-qty-validation

engcom-Hotel commented:
@magento run all tests

engcom-Hotel committed:
Merge branch '2.4-develop' into bug/39479-grouped-product-qty-validation

engcom-Hotel commented:
@magento run all tests

engcom-Hotel reviewed (changes requested):
Hello @Mohamed-Asar,

Thanks for updating the PR as per the review comment!

It seems some tests are still failing due to the changes. Please look into those.

Thanks

Mohamed-Asar committed:
Merge branch '2.4-develop' of https://github.com/Mohamed-Asar/magento2 into bug/39479-grouped-product-qty-validation

Mohamed-Asar committed:
Test case failures fix

Mohamed-Asar commented:
@magento run all tests

Mohamed-Asar commented:
> Hello @Mohamed-Asar,
> 
> Thanks for updating the PR as per the review comment!
> 
> It seems some tests are still failing due to the changes. Please look into those.
> 
> Thanks

@engcom-Hotel Do i need to add the depency in composer.json for this modules( [Magento\InventoryConfigurationApi, Magento\InventorySalesApi]) ? and also these modules are not on this repo, so how can i proceed, do i need to mention in root composer.json and the module composer.json ?

engcom-Hotel commented:
@magento run Static Tests

engcom-Hotel commented:
Hello @Mohamed-Asar,

I guess you need to mention the dependencies in the `composer.json` of `magento/module-catalog-inventory` module here:
https://github.com/Mohamed-Asar/magento2/blob/559d0f8314f93b4aa59562d1a171329242dcffdb/app/code/Magento/CatalogInventory/composer.json#L7-L16

Thanks

Mohamed-Asar commented:
@engcom-Hotel i added before, static test failed due to composer dependency issued, please see this [commit](https://github.com/magento/magento2/pull/39480/commits/baeb177a3bdf99a46a8744ee943f6146c291df9b)

![Screenshot from 2025-04-07 15-56-58](https://github.com/user-attachments/assets/5d397f91-db08-4c15-94c3-a22ab5b14cfd)

engcom-Hotel commented:
@Mohamed-Asar I suggest you to please go through with this documentation: https://developer.adobe.com/commerce/php/architecture/modules/dependencies/

I think we need to add them back.

Thanks

Mohamed-Asar committed:
Merge branch '2.4-develop' of https://github.com/Mohamed-Asar/magento2 into bug/39479-grouped-product-qty-validation

Mohamed-Asar committed:
Composer dependency added

Mohamed-Asar commented:
@magento run all tests

Mohamed-Asar commented:
@engcom-Hotel I’ve added the module dependency in the CatalogInventory module's composer.json, but the static tests are still failing. It seems we need to add the dependency in the root composer.json file. Would it be okay if I go ahead and add it there as well? Please let me know your suggestion.

Also, I noticed that the following [MFTF test](https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Catalog/Test/Mftf/Test/AdminAddInStockProductToTheCartTest.xml) is consistently failing. I reviewed the test case and found that it sets min_qty as 4, but tries to add only 1 quantity to the cart—expecting the product to be added successfully.

However, in this PR, I’ve included min_qty validation, which is why the test is failing. Would it be okay if I update the test case accordingly?

Thanks!

engcom-Hotel commented:
Hello @Mohamed-Asar,

Regarding the MFTF test failure, yes please update the test. As you resolved the bug related to validation which was not in place previously it might happen some existing tests fails. So generally we need to fix those.

And regarding the composer dependency static test failure we will get back to you soon on this.

Thanks

Mohamed-Asar committed:
Mftf test case fix for min qty validation

Mohamed-Asar commented:
@magento run all tests

Mohamed-Asar commented:
Hi @engcom-Hotel

I fixed the mftf test case for min qty validation

Mohamed-Asar committed:
Merge branch '2.4-develop' of https://github.com/Mohamed-Asar/magento2 into bug/39479-grouped-product-qty-validation

Mohamed-Asar committed:
Reverted the msi modules approach

Mohamed-Asar commented:
@magento run Functional Tests B2B, Functional Tests CE, Integration Tests, Static Test

Mohamed-Asar commented:
@magento run Static Tests, Functional Tests B2B

engcom-Hotel commented:
@magento run Functional Tests B2B, Static Tests

engcom-Hotel committed:
Merge branch '2.4-develop' into bug/39479-grouped-product-qty-validation

engcom-Hotel commented:
@magento run all tests

engcom-Hotel committed:
Merge branch '2.4-develop' into bug/39479-grouped-product-qty-validation

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented on app/code/Magento/CatalogInventory/Test/Unit/Model/Product/QuantityValidatorTest.php:
```diff
@@ -0,0 +1,72 @@
+/**
+ * @covers \Magento\CatalogInventory\Model\Product\QuantityValidator
+ */
+class QuantityValidatorTest extends TestCase
```
Covering some edge case tests to ensure the code behaves correctly under unusual or extreme conditions. This could include scenarios like:

- `getMinSaleQty`, `getMaxSaleQty`, or `getQtyIncrements` returning null or invalid values.
- Testing with a `productId` or `websiteId` that does not exist in the stock registry.

engcom-Hotel reviewed (changes requested):
Hello @Mohamed-Asar,
Thanks for the contribution!

Please have a look into the below review comment and fix the failed tests.

Thanks

engcom-Hotel commented:
Yes please @Mohamed-Asar, let us know of you are facing any difficulty in resolving the same.

Mohamed-Asar committed:
Code review changes

Mohamed-Asar commented:
@magento run all tests

Mohamed-Asar commented:
> Hello @Mohamed-Asar, Thanks for the contribution!
> 
> Please have a look into the below review comment and fix the failed tests.
> 
> Thanks

@engcom-Hotel  If i added null i'm getting this error thats why i reverted it 

![image](https://github.com/user-attachments/assets/76e0ac80-5e90-498e-8c75-4ef2fc0568ae)

Mohamed-Asar commented:
> Hello @Mohamed-Asar,
> 
> If you are facing issue in resolving the static test failures, then we can check the possibilities of creating JIRA for the approval.
> 
> Please let us know.
> 
> Thanks

@engcom-Hotel  Let me check and provide you an update by Monday

Mohamed-Asar committed:
Null assigment removed

Mohamed-Asar commented:
@magento run all tests

engcom-Hotel committed:
Merge branch '2.4-develop' into bug/39479-grouped-product-qty-validation

engcom-Hotel commented:
@magento run all tests

engcom-Bravo commented:
Hi @Mohamed-Asar,

Thanks for your Contribution!!.

Could you please resolve the Conflicts to proceed further.

Thanks.

Mohamed-Asar committed:
Conflicts fixed

Mohamed-Asar commented:
@engcom-Bravo I fixed the conflicts, please check

engcom-Hotel commented:
@magento run all tests

engcom-Dash committed:
Merge branch '2.4-develop' into bug/39479-grouped-product-qty-validation

engcom-Dash commented:
@magento run all tests

engcom-Dash commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Static Tests, Unit Tests

engcom-Dash commented:
The consistent test failures for Functional B2B are known Issues and JIRA is raised for them.
Build 1:  https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39480/9ca147ac593cebfa8bb15cddfc34b291/Functional/allure-report-b2b/index.html#categories/8fb3a91ba5aaf9de24cc8a92edc82b5d

![image](https://github.com/user-attachments/assets/1396a319-bd8d-4ef7-8d65-bfd2f66a0bb4)

Build 2: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39480/24ce85251ad7e859543698323eef8b45/Functional/allure-report-b2b/index.html#categories/8fb3a91ba5aaf9de24cc8a92edc82b5d

![image](https://github.com/user-attachments/assets/564b17ed-0273-430a-be9d-7958bc752630)

Known Issues:
EnablePaypalExpressCheckoutAndSubmitAnOrderUsingPaypalExpressCheckoutTest : [ACQE-8007](https://jira.corp.adobe.com/browse/ACQE-8007)
StoreFrontSimpleProductWithSpecialAndTierDiscountPriceTest : [ACQE-7971](https://jira.corp.adobe.com/browse/ACQE-7971)
OrderWithMultipleConfigurableProductsAdditionalStockTest : [ACQE-8060](https://jira.corp.adobe.com/browse/ACQE-8060)

The consistent test failures for Functional CE is known Issues and JIRA is raised for the same.

Build 1: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39480/ea2a8bbe902a91cd403d34cc0f1413f7/Functional/allure-report-ce/index.html#categories/8fb3a91ba5aaf9de24cc8a92edc82b5d

![image](https://github.com/user-attachments/assets/80d6e5d1-9bf3-437c-bb5f-76cd57f482b4)

Build 2: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39480/c6141d1ffdcb2b88bd9059b81502ebd3/Functional/allure-report-ce/index.html#categories/8fb3a91ba5aaf9de24cc8a92edc82b5d

![image](https://github.com/user-attachments/assets/d265228b-f591-4992-97cb-d5bdb66f7468)

Known Issues:
EnablePaypalExpressCheckoutAndSubmitAnOrderUsingPaypalExpressCheckoutTest : [ACQE-8007](https://jira.corp.adobe.com/browse/ACQE-8007)


The consistent test failure for Functional EE is known Issues and JIRA is raised for the same. Other failures are inconsistent and seems to be flaky. They neither part of PR nor failing because of the PR changes.

Build 1: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39480/213c9121f0385ecc0ad9b0794b6719f8/Functional/allure-report-ee/index.html#categories/8fb3a91ba5aaf9de24cc8a92edc82b5d

![image](https://github.com/user-attachments/assets/8aad553c-77d3-4ce2-ae6f-6aa169888d39)

Build 2:https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39480/60b1d7f04550acbc39bb8209e303e64c/Functional/allure-report-ee/index.html#categories/34dd2aa3ce3269b0c73667a1e007a61f/7ad1551983068770/

![image](https://github.com/user-attachments/assets/77e3da25-487e-4391-9a3b-0f33ff5e77b0)

Known Issues:
EnablePaypalExpressCheckoutAndSubmitAnOrderUsingPaypalExpressCheckoutTest : [ACQE-8007](https://jira.corp.adobe.com/browse/ACQE-8007)

engcom-Charlie committed:
Merge branch '2.4-develop' into bug/39479-grouped-product-qty-validation

engcom-Charlie committed:
Merge branch '2.4-develop' into bug/39479-grouped-product-qty-validation
